Bigtable/src/ResumableStream.php

```diff
@@ -220,4 +229,32 @@ private function getMaxRetries(array $options): int
 
         return $retrySettings['maxRetries'] ?? ResumableStream::DEFAULT_MAX_RETRIES;
     }
+
+    private function getTotalTimeoutMillis(array $options): ?int
+    {
+        $retrySettings = $options['retrySettings'] ?? [];
+
+        if ($retrySettings instanceof RetrySettings) {
+            return $retrySettings->getTotalTimeoutMillis();
+        }
+
+        return $retrySettings['totalTimeoutMillis'] ?? 0;
+    }
+
+    private function checkTotalTimeout(int $startTimeMillis): void
+    {
+        if ($this->totalTimeoutMillis > 0) {
+            $elapsedTimeMillis = floor(microtime(true) * 1000) - $startTimeMillis;
+            if ($elapsedTimeMillis > $this->totalTimeoutMillis) {
+                throw new ApiException('Operation timeout exceeeded ', Code::DEADLINE_EXCEEDED);
+            }
+            $remainingTimeMillis = $this->totalTimeoutMillis - $elapsedTimeMillis;
+            // Set timeoutMillis if it's unset or less than the remaining time
+            // in order to preempt requests from exceeding total timeout
+            if (!isset($this->callOptions['timeoutMillis'])
```

Comment:
this is the attempt timeout right?

Reply:
Yes that's correct. `timeoutMillis` is the attempt timeout which is sent to gRPC, and preempts the request if it exceeds that number.